### PR TITLE
(SIMP-4828) Run puppet generate types when needed

### PR DESCRIPTION
**Pull request by trevor-vaughan**
- Added 'pupmod::master::generate_types' which:
  - Adds a script, /usr/local/sbin/simp_generate_types that will run
    'puppet generate types' either on all environments (when first
    distributed by Puppet) or per environment
  - Adds incron hooks to run 'simp_generate_types' when Ruby files in
    any puppet type have been updated
  - Adds incron hooks to run 'simp_generate_types' on any new
    environment
  - Adds incron hooks to run 'simp_generate_types' on *all* environments
    if the puppetserver binary is updated
  - All errors are logged to syslog
- Changed some 'validate_re' statements to 'assert_type' to remove
  deprecation warnings

SIMP-4828 #close

### CODE REVIEW

File: templates/usr/local/sbin/simp_generate_types.epp [hunk 1]
```diff
@@ -0,0 +1,104 @@
+#!/opt/puppetlabs/puppet/bin/ruby
```
Comment (jeefberkey) [inline templates/usr/local/sbin/simp_generate_types.epp]: This looks like a plain file, which is ok to leave as a template

File: spec/spec_helper_acceptance.rb [hunk 2]
```diff
@@ -29,11 +29,43 @@
       # Install modules and dependencies from spec/fixtures/modules
       copy_fixture_modules_to( hosts )
 
+      # Make sure that the SIMP default environment files are in place if they
```
Comment (jeefberkey) [inline spec/spec_helper_acceptance.rb]: Is this stuff related to testing environment leakage?

File: spec/acceptance/suites/default/05_generate_types_spec.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
+require 'spec_helper_acceptance'
+
+describe 'incron driven puppet generate types'  do
+  hosts_with_role(hosts, 'master').each do |host|
+    context "on #{host}" do
+      let(:environment_path) { host.puppet[:environmentpath] }
+      let(:incron_cache) {
+        "#{environment_path}/production/.resource_types"
```
Comment (jeefberkey) [inline spec/acceptance/suites/default/05_generate_types_spec.rb]: Where does this file get generated? Can we name it to something that implies it's part of an incron table?

File: templates/usr/local/sbin/simp_generate_types.epp (hunk 1 above)
Comment (trevor-vaughan) [reply templates/usr/local/sbin/simp_generate_types.epp]: That's mainly for working around masterless scenarios.

File: spec/acceptance/suites/default/05_generate_types_spec.rb (hunk 3 above)
Comment (trevor-vaughan) [reply spec/acceptance/suites/default/05_generate_types_spec.rb]: This is what `puppet generate types` generates. It cannot be changed.

Comment (jeefberkey) [reply spec/acceptance/suites/default/05_generate_types_spec.rb]: :+1:

**Comment by trevor-vaughan** [comment]
Travis tests will continue to fail until https://github.com/simp/pupmod-simp-incron/pull/17 is resolved

File: manifests/master/generate_types.pp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    exec { 'simp_generate_types':
+      command     => "${_generate_types_path} -p all",
+      logoutput   => true,
+      refreshonly => true
+    }
+
+    incron::system_table { 'simp_generate_types':
+      custom_content => epp(
```
Comment (jeefberkey) [inline manifests/master/generate_types.pp]: Why write a new feature in `incron`, then not use it here?

Comment (trevor-vaughan) [reply manifests/master/generate_types.pp]: It was just to not repeat the logic that's over in `incron::system_table`.

File: manifests/init.pp [hunk 5]
```diff
@@ -174,11 +174,12 @@
...
     # These regexes match absolute paths or paths that begin with an existing
     # puppet configuration variable, like $vardir
-    validate_re($classfile,'^(\$(?!/)|/).+')
-    validate_re($confdir,'^(\$(?!/)|/).+')
-    validate_re($environmentpath,'^(\$(?!/)|/).+')
-    validate_re($logdir,'^(\$(?!/)|/).+')
-    validate_re($rundir,'^(\$(?!/)|/).+')
+
+    assert_type(Pattern['^(\$(?!/)|/).+'], $classfile)
+    assert_type(Pattern['^(\$(?!/)|/).+'], $confdir)
```
Comment (lnemsick-simp) [inline manifests/init.pp]: $confdir, $logdir, and $rundir are typed to be Stdlib::Absolutepath.  So puppet configuration variables will not be allowed and these regexes are unnecessary.  Does this mean the types above are incorrect?

File: templates/usr/local/sbin/simp_generate_types.epp [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+require 'ostruct'
+require 'optparse'
+require 'syslog/logger'
+
+log = Syslog::Logger.new(File.basename(__FILE__))
```
Comment (lnemsick-simp) [inline templates/usr/local/sbin/simp_generate_types.epp]: Should you set the syslog facility?

File: templates/usr/local/sbin/simp_generate_types.epp [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+fh = File.open(options.logfile, 'w') if options.logfile
+
+unless options.path
+  msg = "Error: You must pass a path"
+
+  $stderr.puts msg
+  $stderr.puts options.help
+
+  log.error(msg) if options.syslog
```
Comment (lnemsick-simp) [inline templates/usr/local/sbin/simp_generate_types.epp]: this repeated log code lends itself to a log method that does the correct thing, if the body of the script was encapsulated into a class that stored the options in an instance variable

File: manifests/master/generate_types.pp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+#   * The string ``PUPPET_ENVIRONMENTPATH`` will be subsituted with all known
+#   * Puppet Environment Paths
+#
+class pupmod::master::generate_types (
+  Boolean                     $enable                         = true,
+  Boolean                     $trigger_on_puppetserver_update = true,
+  Boolean                     $trigger_on_puppet_update       = true,
+  Array[Stdlib::AbsolutePath] $trigger_paths                  = pupmod::generate_types_munge([
+                                '/PUPPET_ENVIRONMENTPATH',
+                                '/PUPPET_ENVIRONMENTPATH/*/modules/*/lib/puppet/type/*.rb'
```
Comment (lnemsick-simp) [inline manifests/master/generate_types.pp]: IN_NO_LOOP would be more effective if it was operating on directories, not individual files.

File: templates/simp_generate_types_incron_rules.epp [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+<% if $trigger_on_puppetserver_update { -%>
+/opt/puppetlabs/server/apps/puppetserver/bin IN_MODIFY,IN_NO_LOOP <%= $simp_generate_types %> -p all -s
+<% } -%>
+<% if $trigger_on_puppet_update { -%>
+/opt/puppetlabs/puppet/bin/puppet IN_MODIFY,IN_NO_LOOP <%= $simp_generate_types %> -p all -s
+<% } -%>
+<%
+  $trigger_paths.each |Stdlib::AbsolutePath $path| {
+-%>
+<%= $path %> IN_MODIFY,IN_CREATE,IN_NO_LOOP <%= $simp_generate_types %> -p $@/$# -s
```
Comment (lnemsick-simp) [inline templates/simp_generate_types_incron_rules.epp]: Does IN_NO_LOOP disable monitoring of ALL events, only events specified in a user's incrontab, or only events related to the file specified?  If the last option, it doesn't seem like it will be real useful.

I do not see anything that guarantees all the files for a type are laid down before the puppet types generation is called.

Comment (trevor-vaughan) [reply templates/simp_generate_types_incron_rules.epp]: `IN_NO_LOOP` disables monitoring of all events in that incrontab.

Being unable to see into the future, I'm adding a `delay` option that will be set to `30 seconds` by default. I've run `puppet generate types` in four sessions simultaneously and have not seen any adverse reactions so other scripts calling the procedure should not cause any issues.

File: manifests/master/generate_types.pp (hunk 8 above)
Comment (trevor-vaughan) [reply manifests/master/generate_types.pp]: Modified this to handle the `types` directory and to also monitor `modules` in case new modules are added.

File: templates/usr/local/sbin/simp_generate_types.epp (hunk 7 above)
Comment (trevor-vaughan) [reply templates/usr/local/sbin/simp_generate_types.epp]: Will keep this in mind for round 2

File: templates/usr/local/sbin/simp_generate_types.epp (hunk 6 above)
Comment (trevor-vaughan) [reply templates/usr/local/sbin/simp_generate_types.epp]: This works as expected and the facility will be inherited by whatever is calling it.

File: manifests/init.pp (hunk 5 above)
Comment (trevor-vaughan) [reply manifests/init.pp]: Removed the asserts that are now obsolete

File: templates/usr/local/sbin/simp_generate_types.epp (hunk 6 above)
Comment (lnemsick-simp) [reply templates/usr/local/sbin/simp_generate_types.epp]: Don't understand 'facility will be inherited'. I understand that Syslog::Logger is a singleton.  However, whenever I have used Syslog::Logger I have explicitly set the facility and not relied on a default value.

File: templates/usr/local/sbin/simp_generate_types.epp [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+current_umask = File.umask
+File.umask(0022)
+
+delay = options.delay/options.target_environments.count.to_f
+
+options.target_environments.each do |env|
+  sleep(delay)
+
+  # Prevent running on the same environment simultaneously
+  next unless %{ps --no-headers -o args -wwC #{puppet_cmd}}.
```
Comment (jeannegreulich) [inline templates/usr/local/sbin/simp_generate_types.epp]: I don't understand why this does not need a %x?
